src/VisualStudio/Core/Def/Implementation/DiagnosticCache/VisualStudioDiagnosticCacheServiceFactory.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Experiments;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.DiagnosticCache
+{
+    [ExportWorkspaceServiceFactory(typeof(IDiagnosticCacheService), ServiceLayer.Host), Shared]
+    internal sealed class VisualStudioDiagnosticCacheServiceFactory : IWorkspaceServiceFactory
+    {
+        private readonly IDiagnosticService _diagnosticService;
+        private readonly IDiagnosticUpdateSourceRegistrationService _registrationService;
+        private readonly IAsynchronousOperationListenerProvider _listenerProvider;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public VisualStudioDiagnosticCacheServiceFactory(
+            IDiagnosticService diagnosticService,
+            IDiagnosticUpdateSourceRegistrationService registrationService,
+            IAsynchronousOperationListenerProvider listenerProvider)
+        {
+            _diagnosticService = diagnosticService;
+            _registrationService = registrationService;
+            _listenerProvider = listenerProvider;
+        }
+
+        public IWorkspaceService? CreateService(HostWorkspaceServices workspaceServices)
+        {
+            if (workspaceServices.Workspace is not VisualStudioWorkspace vsWorkspace)
+            {
+                return new NoOpDiagnosticCacheService();
+            }
+
+            var experimentationService = workspaceServices.GetRequiredService<IExperimentationService>();
+            if (!experimentationService.IsExperimentEnabled(WellKnownExperimentNames.DiagnosticCache))
```

Comment:
📝 Experiments need to be backed by both a registry option and an experiment name, otherwise it is impossible to enable the feature even for testing on systems with telemetry disabled. See `OOPServerGC` for examples.

Reply:
Added here: https://github.com/dotnet/roslyn/blob/8b11700351e478dfc630e141340d9d23f6d74b8a/src/VisualStudio/Core/Def/Implementation/DiagnosticCache/VisualStudioDiagnosticCacheServiceFactory.cs#L41

How do I enable it? This command doesn't seem to work
```
vsregedit set local HKCU "Roslyn\Internal\OnOff\Features" DiagnosticCache dword 1
```